File: app/src/main/java/com/arejas/dashboardofthings/data/sources/database/converters/DotTypeConverters.java
package com.arejas.dashboardofthings.data.sources.database.converters;

import androidx.room.TypeConverter;

import com.arejas.dashboardofthings.utils.Enumerators;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.Date;
import java.util.Map;

public class DotTypeConverters {

    @TypeConverter
    public static Enumerators.ElementType valueOfElementType(int ordinal) {
        return Enumerators.ElementType.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntElementType(Enumerators.ElementType value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Enumerators.NetworkType valueOfNetworkType(int ordinal) {
        return Enumerators.NetworkType.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntNetworkType(Enumerators.NetworkType value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Enumerators.HttpAuthenticationType valueOfHttpAuthenticationType(int ordinal) {
        return Enumerators.HttpAuthenticationType.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntHttpAuthenticationType(Enumerators.HttpAuthenticationType value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Enumerators.MqttAuthenticationType valueOfMqttAuthenticationType(int ordinal) {
        return Enumerators.MqttAuthenticationType.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntMqttAuthenticationType(Enumerators.MqttAuthenticationType value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Enumerators.DataType valueOfDataType(int ordinal) {
        return Enumerators.DataType.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntDataType(Enumerators.DataType value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Enumerators.MessageType valueOfMessageType(int ordinal) {
        return Enumerators.MessageType.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntMessageType(Enumerators.MessageType value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Enumerators.MqttQosLevel valueOfMqttQosLevel(int ordinal) {
        return Enumerators.MqttQosLevel.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntMqttQosLevel(Enumerators.MqttQosLevel value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Enumerators.HttpMethod valueOfHttpCommand(int ordinal) {
        return Enumerators.HttpMethod.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntLogLevel(Enumerators.LogLevel value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Enumerators.LogLevel valueOfLogLevel(int ordinal) {
        return Enumerators.LogLevel.valueOf(ordinal);
    }

    @TypeConverter
    public static Integer getValueIntHttpCommand(Enumerators.HttpMethod value) {
        return value.ordinal();
    }

    @TypeConverter
    public static Map<String, String> valueofMapStringString(String value) {
        Type mapType = new TypeToken<Map<String, String>>() {
        }.getType();
        return new Gson().fromJson(value, mapType);
    }

    @TypeConverter
    public static String getValueStringMapStringString(Map<String, String> map) {
        Gson gson = new Gson();
        return gson.toJson(map);
    }
    @TypeConverter
    public static Date valueofDate (Long value) {
        return value == null ? null : new Date(value);
    }

    @TypeConverter
    public static Long getValueLongDate (Date date) {
        return date == null ? null : date.getTime();
    }

}

File: app/src/main/java/com/arejas/dashboardofthings/data/sources/database/converters/DotTypeConverters.kt
package com.arejas.dashboardofthings.data.sources.database.converters

import androidx.room.TypeConverter

import com.arejas.dashboardofthings.utils.Enumerators
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import java.lang.reflect.Type
import java.util.Date

object DotTypeConverters {

    @TypeConverter
    fun valueOfElementType(ordinal: Int): Enumerators.ElementType {
        return Enumerators.ElementType.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntElementType(value: Enumerators.ElementType): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueOfNetworkType(ordinal: Int): Enumerators.NetworkType {
        return Enumerators.NetworkType.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntNetworkType(value: Enumerators.NetworkType): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueOfHttpAuthenticationType(ordinal: Int): Enumerators.HttpAuthenticationType {
        return Enumerators.HttpAuthenticationType.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntHttpAuthenticationType(value: Enumerators.HttpAuthenticationType): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueOfMqttAuthenticationType(ordinal: Int): Enumerators.MqttAuthenticationType {
        return Enumerators.MqttAuthenticationType.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntMqttAuthenticationType(value: Enumerators.MqttAuthenticationType): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueOfDataType(ordinal: Int): Enumerators.DataType {
        return Enumerators.DataType.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntDataType(value: Enumerators.DataType): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueOfMessageType(ordinal: Int): Enumerators.MessageType {
        return Enumerators.MessageType.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntMessageType(value: Enumerators.MessageType): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueOfMqttQosLevel(ordinal: Int): Enumerators.MqttQosLevel {
        return Enumerators.MqttQosLevel.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntMqttQosLevel(value: Enumerators.MqttQosLevel): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueOfHttpCommand(ordinal: Int): Enumerators.HttpMethod {
        return Enumerators.HttpMethod.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntLogLevel(value: Enumerators.LogLevel): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueOfLogLevel(ordinal: Int): Enumerators.LogLevel {
        return Enumerators.LogLevel.valueOf(ordinal)
    }

    @TypeConverter
    fun getValueIntHttpCommand(value: Enumerators.HttpMethod): Int {
        return value.ordinal
    }

    @TypeConverter
    fun valueofMapStringString(value: String): Map<String, String>? {
        val mapType = object : TypeToken<Map<String, String>>() {

        }.type
        return Gson().fromJson<Map<String, String>>(value, mapType)
    }

    @TypeConverter
    fun getValueStringMapStringString(map: Map<String, String>): String {
        val gson = Gson()
        return gson.toJson(map)
    }

    @TypeConverter
    fun valueofDate(value: Long?): Date? {
        return if (value == null) null else Date(value)
    }

    @TypeConverter
    fun getValueLongDate(date: Date?): Long? {
        return date?.time
    }

}
